src/notebooks/controllers/notebookControllerManager.ts

```diff
@@ -457,7 +473,27 @@ export class NotebookControllerManager implements INotebookControllerManager, IE
             }
         });
 
-        return defaultPython3Kernel || defaultPythonKernel || defaultPythonLanguageKernel;
+        const defaultController = defaultPython3Kernel || defaultPythonKernel || defaultPythonLanguageKernel;
+
+        if (language === PYTHON_LANGUAGE) {
+            return defaultController;
+        } else {
+            let matchingKernelNameController: VSCodeNotebookController | undefined;
+            let matchingKernelLanguageController: VSCodeNotebookController | undefined;
+            controllers.forEach((item) => {
+                // Sort out interactive or non-interactive controllers
+                if (item.connection.kind !== 'startUsingRemoteKernelSpec') {
+                    return;
+                }
+                if (item.connection.kernelSpec.name === kernelName) {
+                    matchingKernelNameController = item;
+                } else if (item.connection.kernelSpec.language === language) {
```

Comment:
When I moved over the exact match code I ran into one existing error case where the kernelspec language didn't match up with the notebook language_info languages due to capitalization, so maybe need a toLower check here? Pretty sure it was for a C# file.

Reply:
Oh shoot, I thought I had a translation code, will use that.

Reply:
The function that you can use to translate from kenel to monaco language is `translateKernelLanguageToMonaco`
Here it doesn't apply as we're only dealing with languages of kernelspecs & no language components from VS Code (monaco).